csi/moac/volume.ts

```diff
@@ -153,94 +176,83 @@ export class Volume {
     ].indexOf(this.state) < 0) {
       throw new GrpcError(
         GrpcCode.INTERNAL,
-        `Cannot publish "${this}" that isn\'t healthy, degraded or offline`
+        `Cannot publish "${this}" that is neither healthy, degraded nor offline`
       );
     }
 
-    // FSA will take care of publishing the nexus
+    let uri = this.nexus && this.nexus.getUri();
+
     let nexusNode = this._desiredNexusNode(this._activeReplicas(), nodeId);
     if (!nexusNode) {
       // If we get here it means that nexus is supposed to be already published
       // but on a node that is not part of the cluster (has been deregistered).
-      let uri = this.nexus && this.nexus.getUri();
       if (!uri) {
         throw new GrpcError(
           GrpcCode.INTERNAL,
           `Cannot publish "${this}" because the node does not exist`
         );
       }
       return uri;
+    // If the publish has been already done on the desired node then return.
+    } else if (uri && this.nexus?.node?.name === nexusNode.name) {
+      return uri;
     }
+
+    // Set the new desired state
     this.publishedOn = nexusNode.name;
-    this.fsa();
 
-    let self = this;
-    function waitForPublish(resolve: (val: string) => void, reject: (err: any) => void) {
-      if (!self.publishedOn) {
-        // Before the operation could complete the nexus was unpublished.
-        return reject(new GrpcError(
-          GrpcCode.INTERNAL,
-          'The publish operation was cancelled'
-        ));
-      }
-      if (self.nexus?.node?.name === self.publishedOn) {
-        let uri = self.nexus.getUri();
-        if (uri) {
-          log.info(`Published "${self}" at ${uri}`);
-          return resolve(uri);
-        }
-      }
-      self._onStateChange(waitForPublish.bind(self, resolve, reject));
-    }
-    return new Promise((resolve, reject) => {
-      waitForPublish(resolve, reject);
-    });
+    // Cancel any unpublish that might be in progress
```

Comment:
why do we cancel the unpublish and not a previous publish?

Reply:
there cannot be a previous publish. All CSI calls of given type for given volume are serialized. We cancel unpublish because it might be in progress, stuck or waiting for something.